src/Psalm/Internal/Provider/ReturnTypeProvider/GetObjectVarsReturnTypeProvider.php

```diff
@@ -41,6 +42,14 @@ public static function getGetObjectVarsReturnType(
             $atomics = $first_arg_type->getAtomicTypes();
             $object_type = reset($atomics);
 
+            if ($object_type instanceof TNamedObject
+                && strtolower($object_type->value) === strtolower(stdClass::class)
+                && count($intersection_types = $object_type->getIntersectionTypes() ?: []) === 1
+                && ($inner_object_type = reset($intersection_types)) instanceof TObjectWithProperties
+            ) {
+                $object_type = $inner_object_type;
+            }
```

Comment:
This seems a bit hacky, I think what we should probably do is change how `Config::$universal_object_crates` works so that unknown properties aren't allowed to be read without checking if they exist, but writing them is fine (and should produce `&object{newProperty: [type]`, eg for `$obj = object{foo: string, bar: string}`, `$obj->foobar = 1` results in `object{foo: string, bar: string, foobar: 1}`).

Basically I'd like to see [this](https://github.com/vimeo/psalm/issues/7916#issuecomment-1115435524) fixed (so that [this](https://psalm.dev/r/75644eb2ad) can't happen), but maybe there's something I'm not considering here, @orklah do you know if there's a good reason it works like that?

Reply:
Also, since this would be a bigger change, if we go with it maybe this should target master?

Reply:
I don't think there is a big design decision behind how it works right now. Universal object crates, intersection types (on named object) and the cast from array to object were just three separate modules that happen to work relatively well with each other

Reply:
Seems fair! I'll try to figure out something about this on the next few days :)